wasm/jit/jit_arm64.go

```diff
@@ -19,297 +43,423 @@ func jitcall(codeSegment, engine uintptr)
+	// We can choose arbitrary number instead of 1024 which indicates the cache size in the compiler.
+	// TODO: optimize the number.
+	b, err := asm.NewBuilder("arm64", 1024)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a new assembly builder: %w", err)
+	}
+
+	compiler := &arm64Compiler{
+		f:             f,
+		builder:       b,
+		locationStack: newValueLocationStack(),
+	}
+
+	return compiler, nil
+}
+
+type arm64Compiler struct {
+	builder *asm.Builder
+	f       *wasm.FunctionInstance
+	// locationStack holds the state of wazeroir virtual stack.
+	// and each item is either placed in register or the actual memory stack.
+	locationStack *valueLocationStack
+}
+
+func (c *arm64Compiler) generate() (code []byte, staticData compiledFunctionStaticData, maxStackPointer uint64, err error) {
+	code, err = mmapCodeSegment(c.builder.Assemble())
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (c *arm64Compiler) newInstruction() (inst *obj.Prog) {
+	inst = c.builder.NewProg()
+	return
+}
+
+func (c *arm64Compiler) addInstruction(inst *obj.Prog) {
+	c.builder.AddInstruction(inst)
+}
+
+func (c *arm64Compiler) String() (ret string) { return }
+
+func (c *arm64Compiler) emitPreamble() error {
+	// The assembler skips the first instruction so we intetionally add NOP here.
+	nop := c.newInstruction()
+	nop.As = obj.ANOP
+	c.addInstruction(nop)
+	return nil
+}
+
+func (c *arm64Compiler) returnFunction() {
+
+	// TODO: we don't support function calls yet.
+	// For now the following code just simply returns to Go code.
+
+	// Since we return from the function, we need to decement the callframe stack pointer.
+	callFramePointerReg, _ := c.locationStack.takeFreeRegister(generalPurposeRegisterTypeInt)
+	loadCurrentCallFramPointer := c.newInstruction()
+	loadCurrentCallFramPointer.As = arm64.AMOVD
+	loadCurrentCallFramPointer.To.Type = obj.TYPE_REG
+	loadCurrentCallFramPointer.To.Reg = callFramePointerReg
+	loadCurrentCallFramPointer.From.Type = obj.TYPE_MEM
+	loadCurrentCallFramPointer.From.Reg = reservedRegisterForEngine
+	loadCurrentCallFramPointer.From.Offset = engineGlobalContextCallFrameStackPointerOffset
+	c.addInstruction(loadCurrentCallFramPointer)
+
+	decCallFrameStackPointer := c.newInstruction()
+	decCallFrameStackPointer.As = arm64.ASUBS
+	decCallFrameStackPointer.To.Type = obj.TYPE_REG
+	decCallFrameStackPointer.To.Reg = callFramePointerReg
+	decCallFrameStackPointer.From.Type = obj.TYPE_CONST
+	decCallFrameStackPointer.From.Offset = 1
+	c.addInstruction(decCallFrameStackPointer)
+
+	writeDecrementedCallFrameStackPoitner := c.newInstruction()
+	writeDecrementedCallFrameStackPoitner.As = arm64.AMOVD
+	writeDecrementedCallFrameStackPoitner.To = loadCurrentCallFramPointer.From
+	writeDecrementedCallFrameStackPoitner.From = loadCurrentCallFramPointer.To
+	c.addInstruction(writeDecrementedCallFrameStackPoitner)
+
+	c.exit(jitCallStatusCodeReturned)
+	return
+}
+
+func (c *arm64Compiler) exit(status jitCallStatusCode) {
```

Comment:
Let's not forget to eventually document this, as "exit" could leap a brain into many places (similar in amd64). Also, probably rename to compileExit to remove at least some questions.